File: 41_linked_list_cycle.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        ListNode *l = head;
        ListNode *r = head;

		while (l && r && r->next) {
			l = l->next;
			r = r->next->next;
			if (l == r)
				return true;
		}
		return false;
    }
};

File: 41_linked_list_cycle.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        l = head
        r = head
        
        while l and r:        
            l = l.next
            if not l:
                return False
            if r and r.next and r.next:
                r = r.next.next
            else:
                return False
            if l == r:
                return True
        return False
